packages/next/client/use-intersection.tsx

```diff
@@ -87,10 +91,30 @@ function observe(
   }
 }
 
-const observers = new Map<string, Observer>()
+const observers = new Map<idKey, Observer>()
+
+const idList: idKey[] = []
+function getKeyIndex(x: idKey) {
+  for (let i = 0; i < idList.length; i++) {
+    if (idList[i].root === x.root && idList[i].margin === x.margin) {
+      return i
+    }
+  }
+  return -1
+}
 function createObserver(options: UseIntersectionObserverInit): Observer {
-  const id = options.rootMargin || ''
-  let instance = observers.get(id)
+  const id = {
+    root: options.root || null,
+    margin: options.rootMargin || '',
+  }
+  let index = getKeyIndex(id)
+  let instance
+  if (index > -1) {
+    instance = observers.get(idList[index])
+  } else {
+    instance = observers.get(id)
+    idList.push(id)
```

Comment:
When can we remove from this array? This will grow unbounded as the number of images increases.

Reply:
That was something that  I was going to ask as well! Well let me work on it..

Reply:
When I think about this problem more, the idea is that we need to know if the `root` prop has been seen before at a given `rootMargin`.

So what about if we created nested maps? The first could be a WeakMap to `root` element and then we can nest the `rootMargin`.

```js
// initialize
const observers = new WeakMap<HTMLElement, new Map<string, Observer>>()

// usage
const instance = observers.get(root)?.get(margin)
```

I think the WeakMap will automatically remove the reference when the DOM node is destroyed.